code/datums/helper_datums/input.dm

```diff
@@ -0,0 +1,114 @@
+/proc/input_async(mob/user=usr, var/prompt, var/list/choices)
+	var/datum/async_input/A = new(choices, prompt)
+	A.show(user)
+	return A
+
+/proc/input_ranked_async(mob/user=usr, var/prompt="Order by greatest to least preference", var/list/choices)
+	var/datum/async_input/ranked/A = new(choices, prompt)
+	A.show(user)
+	return A
+
+/datum/async_input
+	var/datum/browser/popup
+	var/list/choices
+	var/flash = TRUE
+	var/immediate_submit = FALSE
+	var/prompt
+	var/result = null
+	var/style = "text-align: center;"
+	var/window_id
+	var/height = 200
+	var/width = 400
+
+/datum/async_input/New(var/list/choices, var/prompt="Pick an option:", var/window_id="async_input")
+	src.choices = choices
+	src.prompt = prompt
+	src.window_id = window_id
+
+/datum/async_input/proc/close()
+	if(popup)
+		popup.close()
+	return result
+
+/datum/async_input/proc/show(mob/user)
+	var/dat = create_ui(user)
+	popup = new(user, window_id, , width, height, src)
+	popup.set_content(dat)
+	if(flash && result == null)
+		window_flash(user.client)
+	popup.open()
+
+/datum/async_input/proc/create_ui(mob/user)
+	var/dat = "<div style=\"[style]\">"
+	dat += render_prompt(user)
+	dat += render_choices(user)
+	dat += "<br>"
+	dat += "<br>"
+	dat += button("Submit", "submit=1", , result == null && !immediate_submit)
+	dat += "</div>"
+	return dat
+
+/datum/async_input/proc/render_prompt(mob/user)
+	return "<h2>[prompt]</h2>"
+
+/datum/async_input/proc/render_choices(mob/user)
+	var/dat = " "
+	for(var/choice in choices)
+		dat += button(choice, "choice=[choice]", choice == result)
+		dat += " "
+	return dat
+
+/datum/async_input/proc/button(var/label, var/topic, var/on=FALSE, var/disabled=FALSE)
```

Comment:
no leading `var/` in arg declarations please

Reply:
So that's how it works. Oops 🤦

Reply:
Mm, `var/` in args is always implicit so it's never needed 😄

Reply:
Yeah... for some reason, I thought var was only implicit with inbuilt types like mob/ and client/ 😂